### DISCUSSION TITLE
Add return value to step in report

### DISCUSSION THREAD

**Discussion by kiragengis** [allure-framework/allure2#2352] [Questions & Support] [answered]
I am using Allure with C# and NUnit.

When reading the allure report, I sometimes have to debug just to see the value returned from a step.
Having it printed in the report would be very helpful.

How can I print the return values for all steps? (I would be happy with a default=true, but this might cause some backward compatibility issues)

Is there any step annotation or a parameter that will print the return value to the report?

**Comment by delatrie** [comment]
Hi, @kiragengis!
Allure doesn't have a concept of a step's return value. The closest you can get is to add an extra parameter "returnValue" (or with a similar name), that stores the value you need.
You may define a helper function and call it instead of AllureApi.Step. Something like that:

```csharp
public static T Step<T>(string name, Func<T> func) => AllureApi.Step(name, () =>
{
    var returnValue = func();
    AllureApi.AddParameter("returnValue", returnValue);
    return returnValue;
});
```

**Reply by kiragengis** [reply]
Hi @delatrie,

I tried your solution, but did not find `AllureApi.AddParameter`. I only see `AllureApi.AddTestParameter`. What am I missing? 

This is my function:
```
public static T Step<T>(string name, Func<T> func) => AllureApi.Step(name, () =>
{
    var returnValue = func();
    AllureApi.AddTestParameter("returnValue", returnValue);
    return returnValue;
});
```

And this is how I am using it:
```
[Test]
public void Test1()
{
    Step("My custom step", () => MethodWithCustomParameter(new WhoAmI("string from test", 15)));
}
```

And by the way, with my custom parameter defined as:
```
public class WhoAmI
{
    string myString;
    int myInt;

    public WhoAmI(string myString, int myInt)
    {
        this.myString = myString;
        this.myInt = myInt;
    }

    public override string ToString()
    {
        return $"{myString} : {myInt}";
    }
}
```

Allure is not able to printout the conversion to string.

**Reply by kiragengis** [reply]
@delatrie I found an example in the GitHub project to solve the problem of printing out the custom parameter.
I created the class:
```
public class CustomClassFormatter : TypeFormatter<WhoAmI>
{
    public override string Format(WhoAmI value) => value.ToString();
}
```
And I am using it like this:
```
[OneTimeSetUp]
public static void OneTimeSetUp()
{
    AllureLifecycle.Instance.AddTypeFormatter(new CustomClassFormatter());
}
```

Is this the correct approach?

**Reply by delatrie** [reply, marked as answer]
Oops, I've messed it up. Sorry for that.

The only way to add a parameter to a step is to use `AllureLifecycle`'s `UpdateStep`method ([see here](https://allurereport.org/docs/nunit-reference/#test-steps)). Here is the correct version of the above snippet:

```csharp
public static T Step<T>(string name, Func<T> func) => AllureApi.Step(name, () =>
{
    var returnValue = func();
    AllureLifecycle.Instance.UpdateStep(s => s.parameters.Add(new()
    {
        name = "returnValue",
        value = returnValue.ToString()
    }));
    return returnValue;
});
```

As you can see, there is no need for a `TypeFormatter` (as long as you use `ToString` to convert the value). You have used a `TypeFormatter` correctly, but it doesn't affect parameters added via `Update*` functions (they are too low-level).

**Reply by kiragengis** [reply]
Thank you @delatrie 
This worked for me. 

Can I implement my attribute? The current code is:
```
[AllureStep("A private method with Custom object")]
private string MethodWithCustomParameter(WhoAmI test)
{
    
    return "WhoAmIReturning";
}
```
And I would like to do something like:
```
[CustomAllureStep("A private method with Custom object")]
private string MethodWithCustomParameter(WhoAmI test)
{
    
    return "WhoAmIReturning";
}
```

**Reply by baev** [reply]
@kiragengis plz create a separate discussion for this topic
